### PR TITLE
add build multi-platform cmds script to CI

### PR DESCRIPTION
**Pull request by NicolasMahe**
This PR builds the 2 cosmos clis for linux and mac and upload them to the GitHub release.
Like this we can easily use them to run stuff not implemented in the Engine without having to build them locally.

### CODE REVIEW

**Comment by antho1404** [comment]
Already discussed here https://github.com/mesg-foundation/engine/pull/1552#discussion_r361250057 and Github release was not the right place for it. I still don't really like this idea, it's confusing because it looks that the engine is just the CLI then.
We could publish them on s3 or ipfs or something like that. Or otherwise, they could be embedded in the docker image of the engine and in that case be part of the versioning of the engine

File: .circleci/config.yml [hunk 1]
```diff
@@ -95,6 +95,22 @@ jobs:
       - run: docker login -u $DOCKER_USER -p $DOCKER_PASS
       - run: make docker-publish version=$CIRCLE_TAG
 
+  "release_cli_dev":
+    <<: *run_on_docker
+    steps:
+      - checkout
+      - run: make build-cmd-all version="dev build `echo $CIRCLE_SHA1 | cut -c1-7`"
+      - run: go get -u github.com/tcnksm/ghr
+      - run: ghr -u mesg-foundation -r engine -delete -prerelease -n "Developer Release" -b "Warning - this is a developer release, use it only if you know what are doing. Make sure to pull the latest \`mesg/engine:dev\` image." release-dev ./bin
```
Comment (krhubert) [inline .circleci/config.yml]: Please put that as makefile target and let's only call make from CI (if it's possible). Same below with `release_cli_prod`

File: scripts/build-cmd.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash -e	
+
+if [[ -z "$1" ]]; then	
+  echo -e "version is not set, run:\n"	
+  echo "$0 VERSION"
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
  echo "$0 [version]"
```

File: scripts/build-cmd.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash -e	
+
+if [[ -z "$1" ]]; then	
+  echo -e "version is not set, run:\n"	
+  echo "$0 VERSION"	
+  exit 1	
+fi	
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"
```

File: scripts/build-cmd.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+if [[ -z "$1" ]]; then	
+  echo -e "version is not set, run:\n"	
+  echo "$0 VERSION"	
+  exit 1	
+fi	
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"	
+archs=(amd64 386)
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
archs=(amd64 386)
```

File: scripts/build-cmd.sh [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+if [[ -z "$1" ]]; then	
+  echo -e "version is not set, run:\n"	
+  echo "$0 VERSION"	
+  exit 1	
+fi	
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"	
+archs=(amd64 386)	
+oss=(linux darwin)
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
oss=(linux darwin)
```

File: scripts/build-cmd.sh [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"	
+archs=(amd64 386)	
+oss=(linux darwin)	
+
+for os in ${oss[*]}; do	
+  for arch in ${archs[*]}; do	
+    echo "Building $os $arch..."	
+    GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-daemon-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-daemon/
+    GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-cli-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-cli/
+  done
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
  done
```

File: scripts/build-cmd.sh [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  echo "$0 VERSION"	
+  exit 1	
+fi	
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"	
+archs=(amd64 386)	
+oss=(linux darwin)	
+
+for os in ${oss[*]}; do
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
for os in ${oss[*]}; do
```

File: scripts/build-cmd.sh [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+fi	
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"	
+archs=(amd64 386)	
+oss=(linux darwin)	
+
+for os in ${oss[*]}; do	
+  for arch in ${archs[*]}; do	
+    echo "Building $os $arch..."
```
Comment (krhubert) [inline scripts/build-cmd.sh]:
```suggestion
    echo "Building $os $arch..."
```

File: .circleci/config.yml (hunk 1 above)
Comment (krhubert) [reply .circleci/config.yml]: Lookin on CI, we can create `install-go` target in Makefile, but not related for this PR

**Comment by NicolasMahe** [comment]
@antho1404 can you review

File: scripts/publish-cmds.sh [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  done
+done
+
+go install github.com/tcnksm/ghr
+
+if [[ "$2" == "dev" ]]; then
+  ghr -u mesg-foundation -r engine -replace $1 ./bin
+fi
+if [[ "$2" == "prod" ]]; then
+  ghr -u mesg-foundation -r engine -delete -prerelease -n "Developer Release" -b "Warning - this is a developer release, use it only if you know what are doing. Make sure to pull the latest \`mesg/engine:dev\` image." release-dev ./bin
```
Comment (antho1404) [inline scripts/publish-cmds.sh]: If prod then maybe we need to remove this message, or at least change the docker image to pull

File: scripts/publish-cmds.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+if [[ -z "$1" ]]; then
+  echo -e "version is not set, run:\n"
+  echo "$0 [version] [prod|dev]"
+  exit 1
+fi
+
+if [[ -z "$2" ]]; then
+  echo -e "release type is not set, run:\n"
+  echo "$0 [version] [prod|dev]"
```
Comment (antho1404) [inline scripts/publish-cmds.sh]: wrong copy/paste here I guess.
I'm not so sure to understand the difference between the 2 args. Shouldn't the version be version of the docker image so latest/v0.20...

File: scripts/publish-cmds.sh [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"
+archs=(amd64 386)
+oss=(linux darwin)
+
+for os in ${oss[*]}; do
+  for arch in ${archs[*]}; do
+    echo "Building $os $arch..."
+    GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-daemon-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-daemon/
+    GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-cli-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-cli/
```
Comment (NicolasMahe) [inline scripts/publish-cmds.sh]: let's disabled CGO to be sure the cross-compile works:
```suggestion
    CGO_ENABLED=0 GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-daemon-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-daemon/
    CGO_ENABLED=0 GOOS=$os GOARCH=$arch go build -mod=readonly -o "./bin/mesg-cli-$os-$arch" -ldflags="$LDFLAGS" ./cmd/mesg-cli/
```

File: scripts/publish-cmds.sh [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+fi
+
+if [[ -z "$2" ]]; then
+  echo -e "release type is not set, run:\n"
+  echo "$0 [version] [prod|dev]"
+  exit 1
+fi
+
+
+LDFLAGS="-X 'github.com/mesg-foundation/engine/version.Version=$1'"
```
Comment (NicolasMahe) [inline scripts/publish-cmds.sh]:
```suggestion
LDFLAGS="-s -w -X 'github.com/mesg-foundation/engine/version.Version=$1'"
```

File: scripts/publish-cmds.sh (hunk 9 above)
Comment (NicolasMahe) [reply scripts/publish-cmds.sh]: good catch, I inverted the dev and prod command

File: scripts/publish-cmds.sh (hunk 10 above)
Comment (NicolasMahe) [reply scripts/publish-cmds.sh]: the first argument is to specific the version (eg `v0.19`) and the second is to specify "where" to release, either dev or prod.

File: scripts/publish-cmds.sh [hunk 13]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash -e
+
+if [[ -z "$1" ]]; then
+  echo -e "version is not set, run:\n"
+  echo "$0 [version] [prod|dev]"
```
Comment (antho1404) [inline scripts/publish-cmds.sh]: This comment confused me
```suggestion
  echo "$0 [version] [v0.x.x]"
```

File: scripts/publish-cmds.sh (hunk 10 above)
Comment (antho1404) [reply scripts/publish-cmds.sh]:
```suggestion
  echo "$0 [release type] [prod|dev]"
```

**Pull request merged**